File: web/src/lib/sse.ts
/** Incremental SSE parser (PU4-1 — mirror of U3 sse.py).
 *
 * Push raw chunks in any framing; get complete events out. Tolerant by
 * design: comments and unknown fields ignored, garbage never throws, multiple
 * `data:` lines join with `\n`. A trailing block without its terminator
 * (stream cut mid-event) is never emitted — partial events must not render
 * as if complete.
 */

export interface SseEvent {
  event: string
  data: string
}

const DEFAULT_EVENT = 'message'

function parseBlock(block: string): SseEvent | null {
  let event = DEFAULT_EVENT
  const dataLines: string[] = []
  for (const line of block.split('\n')) {
    if (!line || line.startsWith(':')) continue // comment / keepalive
    const sep = line.indexOf(':')
    const field = sep === -1 ? line : line.slice(0, sep)
    let value = sep === -1 ? '' : line.slice(sep + 1)
    if (value.startsWith(' ')) value = value.slice(1)
    if (field === 'event' && value) event = value
    else if (field === 'data') dataLines.push(value)
  }
  if (dataLines.length === 0 && event === DEFAULT_EVENT) return null
  return { event, data: dataLines.join('\n') }
}

export interface SseParser {
  push(chunk: Uint8Array | string): SseEvent[]
}

export function createSseParser(): SseParser {
  const decoder = new TextDecoder('utf-8') // streaming decode, lone bytes buffered
  let buffer = ''
  return {
    push(chunk) {
      buffer += typeof chunk === 'string' ? chunk : decoder.decode(chunk, { stream: true })
      buffer = buffer.replace(/\r\n/g, '\n').replace(/\r/g, '\n')
      const events: SseEvent[] = []
      let idx = buffer.indexOf('\n\n')
      while (idx !== -1) {
        const block = buffer.slice(0, idx)
        buffer = buffer.slice(idx + 2)
        const parsed = parseBlock(block)
        if (parsed !== null) events.push(parsed)
        idx = buffer.indexOf('\n\n')
      }
      return events
    },
  }
}

File: caduceus/cli/sse.py
"""Minimal SSE parser (tech-stack decision: hand-rolled, ~30 lines of logic).

Consumes an iterator of raw byte chunks (any chunking) and yields
``SseEvent(event, data)`` per spec block. Tolerant by design (PU3-3): unknown
fields and comment lines are ignored, malformed text never raises, multiple
``data:`` lines are joined with ``\\n``.
"""

from __future__ import annotations

from collections.abc import Iterable, Iterator
from dataclasses import dataclass

DEFAULT_EVENT = "message"


@dataclass(frozen=True)
class SseEvent:
    event: str
    data: str


def _parse_block(block: str) -> SseEvent | None:
    event = DEFAULT_EVENT
    data_lines: list[str] = []
    for line in block.split("\n"):
        if not line or line.startswith(":"):
            continue  # comment / keepalive
        field, _, value = line.partition(":")
        value = value.removeprefix(" ")
        if field == "event" and value:
            event = value
        elif field == "data":
            data_lines.append(value)
    if not data_lines and event == DEFAULT_EVENT:
        return None  # nothing meaningful in this block
    return SseEvent(event=event, data="\n".join(data_lines))


def iter_sse(chunks: Iterable[bytes]) -> Iterator[SseEvent]:
    import codecs

    decoder = codecs.getincrementaldecoder("utf-8")(errors="replace")
    buffer = ""
    for chunk in chunks:
        buffer += decoder.decode(chunk)
        buffer = buffer.replace("\r\n", "\n").replace("\r", "\n")
        while "\n\n" in buffer:
            block, buffer = buffer.split("\n\n", 1)
            parsed = _parse_block(block)
            if parsed is not None:
                yield parsed
    # trailing block without terminator (stream cut mid-event) is dropped by
    # design — a partial event must not be rendered as if complete (U3-REL-2
    # keeps already-yielded output; incomplete data is not fabricated)
